src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonSerializerOptions.Converters.cs

```diff
@@ -366,17 +229,17 @@ private JsonConverter GetConverterFromAttribute(JsonConverterAttribute converter
 
         internal bool TryGetDefaultSimpleConverter(Type typeToConvert, [NotNullWhen(true)] out JsonConverter? converter)
         {
-            if (_serializerContext == null && // For consistency do not return any default converters for
-                                    // options instances linked to a JsonSerializerContext,
-                                    // even if the default converters might have been rooted.
-                s_defaultSimpleConverters != null &&
-                s_defaultSimpleConverters.TryGetValue(typeToConvert, out converter))
+            if (_serializerContext != null)
             {
-                return true;
+                // For consistency do not return any default converters for
+                // options instances linked to a JsonSerializerContext,
+                // even if the default converters might have been rooted.
+                converter = null;
```

Comment:
In general we don't want to intermix reflection and source-gen metadata during a serialization routine. The most pertinent reason to me personally is so that we don't mask/create bugs in source-gen/AOT scenarios due to reflection providing silent fallback functionality, and make related analysis more difficult (e.g. https://github.com/dotnet/runtime/pull/68464).

> we currently consistently do exactly opposite and use different converter logic

Indeed we fallback to default converters in the `options.GetConverter(Type)` API. This was to prevent breaking changes where we don't return a converter (see https://github.com/dotnet/runtime/pull/51695#discussion_r618841501,  https://github.com/dotnet/runtime/pull/51897), and more importantly because source-gen does not use or need this method. If it did, I'd have advocated for making the breaking change in 6.0.

> I can make sure this code doesn't throw if reflection logic is not initialized

Is this not the eventual behavior when this method returns `null`?

Reply:
isn't "intermixing" already contradicting the current logic @eiriktsarpalis describes https://github.com/dotnet/runtime/pull/68496#discussion_r859833346 - that means you can partially use reflection APIs after using one of the unsafe APIs? We should decide one way or another